File: objectModel/Java/src/main/java/com/microsoft/commondatamodel/objectmodel/utilities/StringUtils.java
// Copyright (c) Microsoft Corporation.

package com.microsoft.commondatamodel.objectmodel.utilities;

/**
 * String utilities.
 */
public class StringUtils {

  /**
   * Returns part of the string between the two indexes.
   *
   * @param str String to slice
   * @param startIdx Start index
   * @param endIdx End index
   * @return sliced string
   */
  public static String slice(final String str, final int startIdx, final int endIdx) {
    return str.substring(startIdx, endIdx - startIdx);
  }

  /**
   * Returns part of the string from the start index until the end.
   *
   * @param str String to slice
   * @param startIdx Start index
   * @return sliced string
   */
  public static String slice(final String str, final int startIdx) {
    return str.substring(startIdx);
  }

  /**
   * Returns true if the string is null or empty when trimmed (zero length).
   *
   * @param str String to check
   *
   * @return {@code true} if the string is null or empty when trimmed, {@code false} otherwise.
   */
  public static boolean isNullOrTrimEmpty(final String str) {
    return str == null || str.trim().isEmpty();
  }

  /**
   * Returns true if the string is null or empty ("").
   *
   * @param str String to check
   *
   * @return {@code true} if the string is null or empty (""), {@code false} otherwise.
   */
  public static boolean isNullOrEmpty(final String str) {
    return str == null || str.isEmpty();
  }

  /**
   * Capitalizes first letter of the given string.
   * @param str String to capitalize
   * @return Capitalized string
   */
  public static String capitalize(final String str) {
    if (str == null || isNullOrEmpty(str))
      return str;

    if (str.length() == 1)
      return str.substring(0, 1).toUpperCase();
    else
      return str.substring(0, 1).toUpperCase() + str.substring(1);
  }
}

File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Utilities/StringUtils.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Utilities
{
    using System;

    internal static class StringUtils
    {
        public static string Slice(this string str, int startIdx, int endIdx)
        {
            return str.Substring(startIdx, endIdx - startIdx);
        }

        public static string Slice(this string str, int startIdx)
        {
            return str.Substring(startIdx);
        }

        public static bool EndWithOrdinalIgnoreCase(this string str, string specifiedStr)
        {
            return str.EndsWith(specifiedStr, StringComparison.OrdinalIgnoreCase);
        }

        public static bool EqualsWithOrdinalIgnoreCase(this string strA, string strB)
        {
            return strA.Equals(strB, StringComparison.OrdinalIgnoreCase);
        }

        public static int CompareWithOrdinalIgnoreCase(string strA, string strB)
        {
            return string.Compare(strA, strB, StringComparison.OrdinalIgnoreCase);
        }

        public static bool EqualsWithIgnoreCase(this string strA, string strB)
        {
            return strA.Equals(strB, StringComparison.InvariantCultureIgnoreCase);
        }

        public static bool EqualsWithCase(this string strA, string strB)
        {
            return strA.Equals(strB, StringComparison.InvariantCulture);
        }

        /// <summary>
        /// Capitalizes first letter of the given string.
        /// </summary>
        /// <param name="str">The source string to be capitalized.</param>
        /// <returns>The capitalized string which the first letter is upper-cased.</returns>
        public static string CapitalizeValue(this string str)
        {
            if (string.IsNullOrWhiteSpace(str)) 
                return str;
            return char.ToUpper(str[0]) + (str.Length > 1 ? str.Slice(1) : "");
        }

        /// <summary>
        /// Replaces in the pattern in the source with the value.
        /// </summary>
        /// <param name="source">The source string.</param>
        /// <param name="pattern">A pattern in the format {p}. The code will try to find {p} and {P}.</param>
        /// <param name="value">The value to be replaced instead of the pattern.</param>
        /// <returns></returns>
        public static string Replace(string source, string pattern, string value)
        {
            if (value == null)
            {
                value = "";
            }

            string lowerCasePattern = pattern.ToLower();
            string upperCasePattern = pattern.CapitalizeValue();
            string upperCaseValue = !string.IsNullOrEmpty(value) ? value.CapitalizeValue() : "";

            string result = source.Replace($"{{{lowerCasePattern}}}", value);
            return result.Replace($"{{{upperCasePattern}}}", upperCaseValue);
        }
    }
}
